pkg/scheduler/core/equivalence_cache.go

```diff
@@ -208,15 +206,47 @@ func (ec *EquivalenceCache) InvalidateCachedPredicateItemForPodAdd(pod *v1.Pod,
 	ec.InvalidateCachedPredicateItem(nodeName, invalidPredicates)
 }
 
-// getHashEquivalencePod returns the hash of equivalence pod.
-// 1. equivalenceHash
-// 2. if equivalence hash is valid
-func (ec *EquivalenceCache) getHashEquivalencePod(pod *v1.Pod) (uint64, bool) {
-	equivalencePod := ec.getEquivalencePod(pod)
-	if equivalencePod != nil {
-		hash := fnv.New32a()
-		hashutil.DeepHashObject(hash, equivalencePod)
-		return uint64(hash.Sum32()), true
+// getEquivalenceHash computes a hash of the given pod.
+// The hashing function returns the same value for any two pods that are
+// equivalent from the perspective of scheduling.
+func (ec *EquivalenceCache) getEquivalenceHash(pod *v1.Pod) (uint64, bool) {
+	equivalencePod := getEquivalencePod(pod)
+	hash := fnv.New32a()
+	hashutil.DeepHashObject(hash, equivalencePod)
+	return uint64(hash.Sum32()), true
+}
+
+// equivalencePod is the set of pod attributes which must match for two pods to
+// be considered equivalent for scheduling purposes. For correctness, this must
+// include any Pod field which is used by a FitPredicate.
+//
+// NOTE: For equivalence hash to be formally correct, lists and maps in the
+// equivalencePod should be normalized. (e.g. by sorting them) However, the
+// vast majority of equivalent pod classes are expected to be created from a
+// single pod template, so they will all have the same ordering.
+type equivalencePod struct {
+	Namespace      *string
+	Labels         map[string]string
+	Affinity       *v1.Affinity
+	Containers     []v1.Container // See note about ordering
+	InitContainers []v1.Container // See note about ordering
+	NodeName       *string
+	NodeSelector   map[string]string
+	Tolerations    []v1.Toleration
+	Volumes        []v1.Volume // See note about ordering
+}
+
+// getEquivalencePod returns the equivalencePod for a Pod.
+func getEquivalencePod(pod *v1.Pod) *equivalencePod {
+	return &equivalencePod{
+		Namespace:      &pod.Namespace,
```

Comment:
It seems the namspaced resource is used only in: Vlume related predicates, Service Affinity, Pod affinity/anti-affinity (correct me if I made mistake). 

So do you think it make sense that we calculate hash for `namespace + Affinity` and `namespace + Volume` separately, then in most cases equiv cache can be re-used across namespaces?

Reply:
Service Affinity only influenced by Pod label (for selector). It's fine to leave it in hash calculation.

Reply:
> It seems the namspaced resource is used only in: Volume related predicates, Service Affinity, Pod affinity/anti-affinity (correct me if I made mistake).
So do you think it make sense that we calculate hash for namespace + Affinity and namespace + Volume separately, then in most cases equiv cache can be re-used across namespaces?

I am not sure if I understand. Are you saying that we don't need to have `namespace` here?

Reply:
I removed PVCSet because all PVCVolumeInfo objects are in the Volumes list. Based on my understanding of that struct, it is implied to have the same namespace as the pod. i.e. the same claim name could appear in two different namespaces.


If I understand Harry's other comment, he's saying that we should conditionally hash namespace based on whether Affinity or Volume is present. That logic seems valid to me, although the benefit is probably marginal. I will try making the change, but if it makes the code a lot messier, then I'd prefer we keep it as is.

Reply:
Yes, not sure if it's practical but would like to discuss.

Reply:
> it make more sense if equiv class cache can be applied/reused by pods across all namespaces.

I think your point is valid, but I guess the improvement would not be significant as Jonathan said.

Reply:
Agreed.